src/csg_tree.cpp

```diff
@@ -565,77 +444,239 @@ void CsgOpNode::BatchUnion() const {
+            std::vector<std::shared_ptr<const Manifold::Impl>> impls;
+            for (auto &child : frame->left_children)
+              impls.push_back(child->GetImpl());
+            auto result = BatchBoolean(OpType::Intersect, impls);
+            impl->children_ = {std::make_shared<CsgLeafNode>(result)};
+            break;
+          };
+          case OpType::Subtract:
+            if (frame->left_children.empty()) {
+              // nothing to subtract from, so the result is empty.
+              impl->children_ = {std::make_shared<CsgLeafNode>()};
+            } else {
+              BatchUnion(frame->left_children);
+              if (frame->right_children.empty()) {
+                // nothing to subtract, result equal to the LHS.
+                impl->children_ = {frame->left_children[0]};
+              } else {
+                BatchUnion(frame->right_children);
+                Boolean3 boolean(*frame->left_children[0]->GetImpl(),
+                                 *frame->right_children[0]->GetImpl(),
+                                 OpType::Subtract);
+                Manifold::Impl result = boolean.Result(OpType::Subtract);
+                impl->children_ = {std::make_shared<CsgLeafNode>(
+                    std::make_shared<Manifold::Impl>(result))};
+              }
+            }
+            break;
+        }
+      frame->op_node->cache_ = std::static_pointer_cast<CsgLeafNode>(
+          impl->children_[0]->Transform(frame->op_node->transform_));
+      if (frame->destination != nullptr)
+        frame->destination->push_back(std::static_pointer_cast<CsgLeafNode>(
+            frame->op_node->cache_->Transform(frame->transform)));
+      stack.pop_back();
+    } else {
+      auto add_children = [&](std::shared_ptr<CsgNode> &node, OpType op,
+                              mat3x4 transform, auto *children) {
+        if (node->GetNodeType() == CsgNodeType::Leaf)
+          children->push_back(std::static_pointer_cast<CsgLeafNode>(
+              node->Transform(transform)));
+        else
+          stack.push_back(std::make_shared<CsgStackFrame>(
+              false, op, transform, children,
+              std::static_pointer_cast<const CsgOpNode>(node)));
+      };
+      frame->finalize = true;
+      if (frame->op_node->op_ == OpType::Subtract) {
+        for (size_t i = 0; i < impl->children_.size(); i++)
+          add_children(impl->children_[i], OpType::Add, la::identity,
+                       i == 0 ? &frame->left_children : &frame->right_children);
+      } else {
+        // op_node use_count == 2 because it is both inside one CsgOpNode
+        // and in our stack.
+        bool skipFinalize = frame->destination != nullptr &&
+                            frame->op_node->op_ == frame->parent_op &&
+                            frame->op_node.use_count() <= 2 &&
+                            frame->op_node->impl_.UseCount() == 1;
+        if (skipFinalize) stack.pop_back();
```

Comment:
We just set `frame->finalize = true;` above - is that still true? Or does this skipping refer to its child or parent?

Reply:
The frame is removed here, so we will skip the current frame. Anyway, I can make it explicit.